src/main/java/org/jenkinsci/plugins/workflow/job/console/NewNodeConsoleNote.java

```diff
@@ -0,0 +1,133 @@
+import hudson.console.ConsoleAnnotationDescriptor;
+import hudson.console.ConsoleAnnotator;
+import hudson.console.ConsoleNote;
+import hudson.model.TaskListener;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.util.Iterator;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import org.jenkinsci.plugins.workflow.actions.LabelAction;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.graph.BlockEndNode;
+import org.jenkinsci.plugins.workflow.graph.BlockStartNode;
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.jenkinsci.plugins.workflow.log.LogStorage;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Console line with note printed when a new {@link FlowNode} is added to the graph.
+ * Defines the {@code pipeline-new-node} CSS class and several attributes which may be used to control subsequent behavior:
+ * <ul>
+ * <li>{@code nodeId} for {@link FlowNode#getId}
+ * <li>{@code startId} {@link FlowNode#getId} for {@link BlockStartNode}, else {@link BlockEndNode#getStartNode}, else absent
+ * <li>{@code enclosingId} the immediately enclosing {@link BlockStartNode}, if any
+ * <li>{@code label} for {@link LabelAction} if present
+ * </ul>
+ * @see LogStorage#startStep
+ */
+@Restricted(NoExternalUse.class)
+public class NewNodeConsoleNote extends ConsoleNote<WorkflowRun> {
+
+    /**
+     * Prefix used in metadata lines.
+     */
+    private static final String CONSOLE_NOTE_PREFIX = "[Pipeline] ";
+
+    public static void print(FlowNode node, TaskListener listener) {
+        PrintStream logger = listener.getLogger();
+        synchronized (logger) {
+            try {
+                listener.annotate(new NewNodeConsoleNote(node));
+            } catch (IOException x) {
+                // never mind
+            }
+            logger.println(CONSOLE_NOTE_PREFIX + node.getDisplayFunctionName());
+        }
+    }
+
+    private final @Nonnull String id;
+    private final @CheckForNull String enclosing;
+    private final @CheckForNull String start;
+
+    private NewNodeConsoleNote(FlowNode node) {
+        id = node.getId();
+        if (node instanceof BlockEndNode) {
+            enclosing = null;
+            start = ((BlockEndNode) node).getStartNode().getId();
+        } else {
+            Iterator<BlockStartNode> it = node.iterateEnclosingBlocks().iterator();
+            enclosing = it.hasNext() ? it.next().getId() : null;
+            start = node instanceof BlockStartNode ? node.getId() : null;
+        }
+    }
+
+    @Override
+    public ConsoleAnnotator<?> annotate(WorkflowRun context, MarkupText text, int charPos) {
+        StringBuilder startTag = new StringBuilder("<span class=\"pipeline-new-node\" nodeId=\"").append(id);
```

Comment:
nit: Why not close each attribute immediately (e.g. `new StringBuilder("<span class=\"pipeline-new-node\" nodeId=\"").append(id).append("\"")`) to make it more obvious to readers that they are closed correctly?

Reply:
Merely because that involves more concatenations total.